src/output-json-ftp.c

```diff
@@ -0,0 +1,315 @@
+                }
+
+                where += pos;
+                length -= pos;
+            }
+        }
+
+        json_object_set_new(cjs, "reply", js_resplist);
+        json_object_set_new(cjs, "completion_code", js_respcode_list);
+    }
+
+    if (tx->dyn_port) {
+        json_object_set_new(cjs, "dynamic_port", json_integer(tx->dyn_port));
+    }
+
+    if (tx->command_descriptor->command == FTP_COMMAND_PORT ||
+        tx->command_descriptor->command == FTP_COMMAND_EPRT) {
+        json_object_set_new(cjs, "mode",
+                json_string((char *)(tx->active ? "active" : "passive")));
+    }
+
+    json_object_set_new(cjs, "reply_received",
+            json_string((char *)(tx->done ? "yes" : "no")));
+
+    return cjs;
+}
+
+
+static int JsonFTPLogger(ThreadVars *tv, void *thread_data,
+    const Packet *p, Flow *f, void *state, void *vtx, uint64_t tx_id)
+{
+    SCEnter();
+
+    const char *event_type;
+    if (f->alproto == ALPROTO_FTPDATA) {
+        event_type = "ftp_data";
+    } else {
+        event_type = "ftp";
+    }
+    FTPTransaction *tx = vtx;
+    LogFTPLogThread *thread = thread_data;
+    LogFTPFileCtx *ftp_ctx = thread->ftplog_ctx;
+
+    json_t *js = CreateJSONHeaderWithTxId(p, LOG_DIR_FLOW, event_type, tx_id);
+    if (likely(js)) {
+        JsonAddCommonOptions(&ftp_ctx->cfg, p, f, js);
+        json_t *cjs = NULL;
+        if (f->alproto == ALPROTO_FTPDATA) {
+            cjs = JsonFTPDataAddMetadata(f);
+        } else {
+            cjs = JsonFTPLogCommand(f, tx);
+        }
+
+        if (cjs) {
+            json_object_set_new(js, event_type, cjs);
+        }
+
+        MemBufferReset(thread->buffer);
+        OutputJSONBuffer(js, thread->ftplog_ctx->file_ctx, &thread->buffer);
+
+        json_object_clear(js);
+        json_decref(js);
+    }
+    return TM_ECODE_OK;
+}
+
+static void OutputFTPLogDeInitCtxSub(OutputCtx *output_ctx)
+{
+    LogFTPFileCtx *ftplog_ctx = (LogFTPFileCtx *)output_ctx->data;
+    SCFree(ftplog_ctx);
+    SCFree(output_ctx);
+}
+
+
+static OutputInitResult OutputFTPLogInitSub(ConfNode *conf,
+    OutputCtx *parent_ctx)
+{
+    OutputInitResult result = { NULL, false };
+    OutputJsonCtx *ajt = parent_ctx->data;
+
+    LogFTPFileCtx *ftplog_ctx = SCCalloc(1, sizeof(*ftplog_ctx));
+    if (unlikely(ftplog_ctx == NULL)) {
+        return result;
+    }
+    ftplog_ctx->file_ctx = ajt->file_ctx;
+    ftplog_ctx->cfg = ajt->cfg;
+
+    OutputCtx *output_ctx = SCCalloc(1, sizeof(*output_ctx));
+    if (unlikely(output_ctx == NULL)) {
+        SCFree(ftplog_ctx);
+        return result;
+    }
+    output_ctx->data = ftplog_ctx;
+    output_ctx->DeInit = OutputFTPLogDeInitCtxSub;
+
+    SCLogDebug("FTP log sub-module initialized.");
+
+    AppLayerParserRegisterLogger(IPPROTO_TCP, ALPROTO_FTP);
+    AppLayerParserRegisterLogger(IPPROTO_TCP, ALPROTO_FTPDATA);
+
+    result.ctx = output_ctx;
+    result.ok = true;
+    return result;
+}
+
+#define OUTPUT_BUFFER_SIZE 65535
```

Comment:
Should we rather reuse the same macro from output-json.c by placing it in output-json.h (and maybe renaming int `JSON_OUTPUT_BUFFER_SIZE` ?)

Reply:
Good suggestion; I'll take a look at the details and see if it will work out ok.

Reply:
I'll make this change to each of the output modules handling JSON.